### ISSUE TITLE
Execution Block ID uid://A001/X15a0/X192/Sgr_A_st_ao_03_7M

### ISSUE THREAD

**Issue by ashleythomasbarnes** [ACES-CMZ/reduction_ACES#6]
* [ ] Delivered? - completed
* [ ] Downloaded? (specify where)
* [ ] Weblog Quality Assessment?
* [ ] Imaging: Continuum
* [ ] Imaging: Lines
   * [ ] HCO+
   * [ ] HNCO
   * [ ] H13CN
   * [ ] H13CO+/SiO
   * [ ] Cont 1
   * [ ] Cont 2

**Comment by ashleythomasbarnes** [comment]
# HCO+
![Screenshot 2021-11-15 at 14 20 02](https://user-images.githubusercontent.com/30494539/141788835-2cfd2a75-8258-4c7c-9c70-2a3ab8d6fd13.png)
# HNCO
![Screenshot 2021-11-15 at 14 20 07](https://user-images.githubusercontent.com/30494539/141788847-d7293913-2028-4696-8801-3df578b77064.png)

**Comment by keflavich** [comment]
Probable SiO v=1 J=2-1 maser:
![image](https://user-images.githubusercontent.com/143715/141885671-84d13a44-1a90-4b46-ad85-f69c92671791.png)

**Comment by xinglunju** [comment]
**QA summary & issues**
---
- SPW 20 & 22: band phase calibration solutions? The peak flux of the phase calibrator J1744 at these two spws is lower than that in other spws at the same frequency range (~330 mJy/bm vs. ~380 mJy/bm), suggesting phase de-correlation. The achieved image rms for the phase calibrator is 2-3 times higher than the theoretical value. Inspection of the phase solutions shows that the problem is likely with uid___A002_Xf1ecde_X738b.ms. The warning in task 16 is related: the calibrated visibility flux densities of J1744 in spw 20&22 in uid___A002_Xf1ecde_X738b.ms are factor 1.5 lower than the catalog value (see below).
<img width="1029" alt="Screen Shot 2022-01-07 at 10 02 35 PM" src="https://user-images.githubusercontent.com/3944252/148555315-a9c30b73-b630-4679-a391-ae1c7759977e.png">
- So, uid___A002_Xf1ecde_X738b.ms may need a re-calibration, with special attention to spw 20 & 22, maybe using a different spw phase-up map.

**Continuum**
---
- Looks good. The findcont results can be improved.
- The continuum image of spw 20 has a peak that is not seen in any other spws, and the peak value is twice of those in other spws.
![image](https://user-images.githubusercontent.com/3944252/148553848-502cebcf-0892-417e-8740-b2bf54e872f9.png)
![image](https://user-images.githubusercontent.com/3944252/148553879-3c3e71e1-fcbd-4b00-88e9-539ee677a3d3.png)
Could be a bad choice of line-free channel in findcont that includes some line emission (see below)?
![image](https://user-images.githubusercontent.com/3944252/148553649-26cd97a6-d8c3-486f-aff5-7600152afa0d.png)

**Cubes**
---
- Looks ok.
- tclean stopped to prevent divergence for spw 24, although the image and the line-free moment maps seem ok.
- The spectral line in spw 20 is a bit close to the edge, but may be ok. See plots in Ash's comment above.

**Comment by xinglunju** [comment]
@keflavich Yes I agree. At least we need to double check the calibration and flagging of spws 20 & 22 in this EB, in the end we may simply flag these two spws instead of re-running calibration.

**Comment by keflavich** [comment]
Options:

1. Flag out those two windows in that EB
2. Attempt re-calibration
3. Request re-calibration by ALMA (QA3)

**Comment by thusharags** [comment]
Here is a link to the Liszt paper on the absorption issue:
https://ui.adsabs.harvard.edu/abs/2018A&A...610A..49L/abstract

**Comment by keflavich** [comment]
This calibrator does look pretty weird.  This is spw20 for the bad one (from tsys):
![image](https://user-images.githubusercontent.com/143715/149165867-9308ba1e-5969-4e7c-9a1c-5a0c810f852f.png)

These are good ones:
![image](https://user-images.githubusercontent.com/143715/149165922-5609cf90-3e0d-4dd6-a2d6-97a010856a9e.png)
![image](https://user-images.githubusercontent.com/143715/149165962-31828520-8260-46a8-80ff-adb75e48e89c.png)
![image](https://user-images.githubusercontent.com/143715/149165982-d8bc940e-6180-4f38-aad2-9a9e5dab1ab6.png)

All of these spectra look weird (they have lines in the window), but it's interesting that the last one got the right flux scale, while the first did not.

**Comment by keflavich** [comment]
Also interesting that [`tsysflag`](https://data.rc.ufl.edu/secure/adamginsburg/ACES/weblogs/humanreadable/Sgr_A_st_ao_03_7M/html/t2-4m.html?sidebar=sidebar_stage7&ms=all&subpage=t2-4m_details.html) seems to have handled these lines right, though, which makes it less obvious why that one EB was messed up.

**Comment by thusharags** [comment]
Does anyone know what the threshold is for flagging line channels in the tsys cal table? Wonder whether the rogue  dataste  just missed that threshold by a bit? Unlikely though given how strong these lines are in the narrow window.

**Comment by xinglunju** [comment]
I checked calibrated data, and spw 20 of J1744 does show HCO+ absorption features in all the four EBs. See plots below.
![Screenshot from 2022-01-14 14-51-11](https://user-images.githubusercontent.com/3944252/149465103-c02c315c-1c84-47da-8575-87a3fb859e58.png)
![Screenshot from 2022-01-14 14-51-33](https://user-images.githubusercontent.com/3944252/149465261-fcbaa6d3-5ef5-4d9f-abf9-725e3393aec5.png)
![Screenshot from 2022-01-14 14-51-51](https://user-images.githubusercontent.com/3944252/149465275-68a9d42f-9afe-47f7-bd13-3fe39f1fd1b0.png)
![Screenshot from 2022-01-14 14-52-07](https://user-images.githubusercontent.com/3944252/149465283-b4703e8c-e065-4547-84f2-50fb42a6788d.png)

The phase calibration of spw 20 and 22 of the first EB (X738b) doesn't seem good, with large phase scatters. For a comparison, the last two plots show the calibrated phase for the second EB (X784b), which seems good.
![Screenshot from 2022-01-14 14-57-31](https://user-images.githubusercontent.com/3944252/149465349-7e304109-0c9a-4ffb-b065-2e04f8ef6faf.png)
![Screenshot from 2022-01-14 14-57-42](https://user-images.githubusercontent.com/3944252/149465486-0b14b6e2-b310-486c-b6b1-c504c17d23de.png)
(calibrated phase for the second, 'good' EB)
![Screenshot from 2022-01-14 14-58-03](https://user-images.githubusercontent.com/3944252/149465518-d733b738-da7a-4fba-a221-6d0448e9c621.png)
![Screenshot from 2022-01-14 14-57-56](https://user-images.githubusercontent.com/3944252/149465556-40691c86-697c-4811-b638-cf0ed06b73ec.png)

**The calibrated amplitude of J1744 of X738b is messed up**. The last four scans have clearly wrong amplitudes. This is likely the root of the abnormal flux of J1744 in this EB.
![Screenshot from 2022-01-14 14-56-14](https://user-images.githubusercontent.com/3944252/149465622-5739c546-b823-4864-818b-70e69ccb7508.png)
![Screenshot from 2022-01-14 14-56-23](https://user-images.githubusercontent.com/3944252/149465631-ccd87b40-f60c-48f0-adfc-643d8881b63c.png)

**Comment by ksi85** [comment]
Hi all,

I contacted ESO about the flux issue of the phase calibrator in the first measurement set and they suggested that the problem could lie with the used reference antenna. So they suggest to re-run the calibration with reference antenna CM03 instead of CM10 and then check if the calibration improved. If this is not the case, the DRM and me will investigate further.

**Comment by keflavich** [comment]
Suggested solutions:

- [ ] flag out bad scans
- [ ] flag out bad spws in bad EB
- [ ] flag out bad EB

Origin of problem (discussed on telecon):
 * probably change in atmospheric line particularly affecting these two SPWs

**Comment by keflavich** [comment]
@ksi85 will you contact the helpdesk and ask if they can update the flagcommands?

**Comment by ksi85** [comment]
Yes, I will contact the DRM tomorrow to ask if the flagcommands can be updated accordingly and then report back on what she says.

**Comment by xinglunju** [comment]
@ksi85 thank you! Here is what I did:
In package_path/calibration/uid___A002_Xf1ecde_X738b.flagtemplate.txt (or in SBid_*_auxproducts.tgz, which is a tarball that contains this txt file),
add a new flag:
mode='manual' spw='20,22' scan='8~19' reason='QA3:timegaincal_amplitude_time'

**Comment by ksi85** [comment]
Hi all, this is the answer I got from the DRM about the procedure on if/how to get the data in the archive corrected:

_The only way to request the correction of the results in the archive is through a QA3 procedure via HD. You need to describe the problem and request that the archived data should be corrected. I will then open a PRTSPR ticket with your request and documentation. The assignee will decide if this is needed or not. If it will be accepted, then we will go through QA3.
If you really want these data corrected in the archive, send the request today, because due to the Feb shutdown there is some chance that the QA3 procedure will be accepted. In normal operation, this will not be corrected because the reference spw is not affected._

I am very busy today so I don't think I will get to writing the HD message. Can someone else take care of this if you think that it would be useful to have the corrected data in the archive? Thanks!

**Comment by xinglunju** [comment]
I can submit the ticket. Shall I send it to EU ARC (who are supposed to support this project)?

**Comment by ksi85** [comment]
Thank you! Yes, this would be the right ARC :-)

**Comment by xinglunju** [comment]
Oops... my ALMA science portal/helpdesk account was affiliated with the EA ARC so I could only submit tickets to EA. I just changed the preferred ARC to EU, but still I only see EA options in the helpdesk. As far as I remember, the change of preferred ARCs itself is a ticket that needs to be approved by ARC staff. So probably I cannot do it today... 

If someone else can do it today, that would be great. I will post a draft message here asap with all the details. Otherwise I will try it tomorrow once the change of the preferred ARC is approved.

**Comment by xinglunju** [comment]
It seems that my account has been switched to EU, so I've submitted a ticket:

> Dear ARC staff,
> 
> I am working on the delivered data of the cycle 8 large program 2021.1.00172.L as a co-I, and I noticed that there might be issues in one of the five EBs in the SB Sgr_A_st_ao_03_7M. I wonder if it is possible to make adjustments to the data in the archive and re-ingest the products to the science archive, so that other users can assess the (hopefully) correct data.
> 
> The problem can be illustrated in the attached plot. The calibrated flux in spectral windows 20 and 22 of the phase calibrator J1744 in the EB uid://A002/Xf1ecde/X738b has anomalies: the flux drops to about half of the catalog value after the third scan. This has caused low scores for Tasks 16 and 21 in the weblog. We have attempted to investigate the root of this anomaly (e.g., switching the reference antenna) but got little clue so far. Meanwhile we tried to flag these scans in this particular EB, and found the result to be promising. The images from the pipeline after flagging the bad scans have acceptable RMS, and the scores for Tasks 16 and 21 in the weblog become 1.
> 
> The fix we did is simply add a flagging command in the flagtemplate file in the 'calibration' directory in the delivered package. In our case it is uid___A002_Xf1ecde_X738b.flagtemplate.txt, and we added one line in it:
> 
>> mode='manual' spw='20,22' scan='8~19' reason='QA3:timegaincal_amplitude_time'
> 
> and then packed it into the tarball member.uid___A001_X15a0_X192.hifa_calimage_renorm.auxproducts.tgz.
> 
> Please let me know if what we did looks ok to you. If this is indeed a viable fix, could you replace the data in the archive with this updated flagtemplate and with the bad scans flagged?

**Comment by xinglunju** [comment]
**More updates on the re-calibration**
---
**Flagging bad scans and rerunning calibration**
 - I added a flagging command in uid___A002_Xf1ecde_X738b.flagtemplate.txt under the 'calibration' subdirectory of the delivered package:
> mode='manual' spw='20,22' scan='8~19' reason='QA3:timegaincal_amplitude_time'
 - This flagged all the scans toward the phase calibrator J1744 and the target fields after the first two 'good' scans.
 - The way the pipeline handles the flagtemplate file is a bit tricky: by default, scriptForPI.py always unpacks the *auxproducts.tgz tarball in this subdirectory and reads the flagtemplates therein. So I packed the updated uid___A002_Xf1ecde_X738b.flagtemplate.txt into that tarball (the other way is to comment out all the lines in scriptForPI.py that unpack the tarball, and ask it to directly read the available flagtemplates in the subdirectory).
 - Then I reran the pipeline (not just restore the calibrated products). To do this, one needs to move the 'piperestorescript.py' in the 'script' subdirectory somewhere else, and then start scriptForPI.py. Also there is likely a bug in the source code of the task hif_refant, which I have reported in a HD ticket to ARC. I copied the message here:
```
I am trying to re-run (not simply restore) the pipeline calibration of a delivered package, and I think there is a bug in the code:
In the source code of the task hif_refant, which locates at ~/CASAPATH/pipeline/pipeline/hif/cli/hif_refant.py,
at line #90, the keyword 'refant' should be able to accept a list, because sometimes we supply a list of EBs to this task, and each EB can have different reference antennas, then the keyword 'refant' would be a list of different antenna names (e.g., refant=['CM01', 'CM02', 'CM03']). However, the code only allows for a single string. 
To solve this problem, I changed this part of the source code from:
>'refant': {'type': 'cStr', 'coerce': _coerce.to_str},
to:
>'refant': {'type': 'cStrVec', 'coerce': [_coerce.to_list,_coerce.to_strvec]},
I don't know if this has been solved in newer versions of CASA. It may be helpful to remind users about this bug, if confirmed, because some users might also want to re-run the calibration.
```

**Results after flagging**
- The scores for tasks 16 and 21 now are 1.
- However the score for task 1 drops to 0.3 because "Neither primary or backup flux service could be queried. ASDM values used" (see plot below). But I don't think this is a big problem, because the derived fluxes for the calibrators seem to be ok. Any thoughts?
<img width="540" alt="image" src="https://user-images.githubusercontent.com/3944252/153443399-727216a0-0ff2-4517-8fc4-3b8de9f0ff59.png">

- The fluxes of J1744 from task 16 (now the calibrated/catalog flux ratios are all ~1)
<img width="840" alt="image" src="https://user-images.githubusercontent.com/3944252/153444023-cbce8944-3548-49b7-9cef-3a893451c89b.png">

- Calibrator image info from task 21 (note the green scores, all 1.00)
<img width="796" alt="image" src="https://user-images.githubusercontent.com/3944252/153444189-7f284c5c-7d7e-4426-80c7-9389fa6b303c.png">

- Target image info from task 38 (note the green scores, all 1.00)
<img width="800" alt="image" src="https://user-images.githubusercontent.com/3944252/153444323-e930bced-92ea-46a3-bd81-0bf620a60f97.png">

**Overall assessment**
- The continuum and line images seem good, although there is room to improve (better continuum selection, finer masking strategy for line cubes to avoid divergence, etc.).

**Comment by xinglunju** [comment]
**Update on the re-calibration**
---
The ARC has accepted the QA3 request and will re-run the calibration and replace the data in the archive. The problem is actually with the antenna CM02: it is not necessary to flag everything between scan=8 to 19, just flagging CM02 between scan=8 to 19 will solve the problem. The imaging results do not change much, though.

**Comment by xinglunju** [comment]
@keflavich The QA3 package has been delivered on Feb 16, could you download it to hipergator (if it hasn't been done)?

**Comment by keflavich** [comment]
The products are downloaded and the pipeline is being re-run

**Comment by keflavich** [comment]
TODO: I need to put up the new version of the pipeline weblog post-QA3

**Comment by keflavich** [comment]
The pipeline is fully re-run, I think - all the MSes and splits are present.  I see all the image cubes.

...but you think some EBs are missing?

I see:
https://data.rc.ufl.edu/secure/adamginsburg/ACES/weblogs-reimaging/member.uid___A001_X15a0_X192/pipeline-20220223T160940/html/t1-1.html
* [uid___A002_Xf1ecde_X738b.ms](https://data.rc.ufl.edu/secure/adamginsburg/ACES/weblogs-reimaging/member.uid___A001_X15a0_X192/pipeline-20220223T160940/html/t2-1.html?sidebar=sidebar_uid___A002_Xf1ecde_X738b_ms)
* [uid___A002_Xf1ecde_X784b.ms](https://data.rc.ufl.edu/secure/adamginsburg/ACES/weblogs-reimaging/member.uid___A001_X15a0_X192/pipeline-20220223T160940/html/t2-1.html?sidebar=sidebar_uid___A002_Xf1ecde_X784b_ms)
* [uid___A002_Xf20692_Xd716.ms](https://data.rc.ufl.edu/secure/adamginsburg/ACES/weblogs-reimaging/member.uid___A001_X15a0_X192/pipeline-20220223T160940/html/t2-1.html?sidebar=sidebar_uid___A002_Xf20692_Xd716_ms)
* [uid___A002_Xf20692_X110ee.ms](https://data.rc.ufl.edu/secure/adamginsburg/ACES/weblogs-reimaging/member.uid___A001_X15a0_X192/pipeline-20220223T160940/html/t2-1.html?sidebar=sidebar_uid___A002_Xf20692_X110ee_ms)

and in the directory I see:
```
drwxr-sr-x 28 adamginsburg adamginsburg 4.0K Feb 23 08:38 uid___A002_Xf1ecde_X738b.ms
drwxr-sr-x 28 adamginsburg adamginsburg 4.0K Feb 23 08:41 uid___A002_Xf1ecde_X784b.ms
drwxr-sr-x 28 adamginsburg adamginsburg 4.0K Feb 23 08:44 uid___A002_Xf20692_Xd716.ms
drwxr-sr-x 28 adamginsburg adamginsburg 4.0K Feb 23 08:47 uid___A002_Xf20692_X110ee.ms
```

so I think the pipeline run is all good.

**Comment by xinglunju** [comment]
Yes, this one has all the four EBs, but in the 'By Task' tab of the weblog, it goes as far as '10. hif_makeimlist: Set-up parameters for target aggregate continuum imaging' and then stops.

**Comment by keflavich** [comment]
The lack of weblogs is because the cubes are produced by the second approach to reimaging described here:
https://github.com/ACES-CMZ/reduction_ACES/issues/163#issue-1272295566

**Comment by xinglunju** [comment]
Thanks for looking into this. For such pipeline runs, we need to download the images from Globus and review them, right?

**Comment by keflavich** [comment]
Yes, that's right.  I'll work on integrating quicklooks of these cubes to the github issues.

**Comment by xinglunju** [comment]
Overview of the QA3 data
-----
Imaging looks good. Set to 'done'?
The calibration part of the QA3 pipeline should be in the ALMA archive, which is not accessible here, although it should be good.

Spectral windows
-----
Continuum baseline is fitted with 0th order. Results look good. See spw20 and 22 for examples (from pipeline-20220223T160940)
![image](https://user-images.githubusercontent.com/3944252/175249994-eb9b7355-cd69-49db-8ad0-00dc3c5728d7.png)
![image](https://user-images.githubusercontent.com/3944252/175250530-7e6e8736-3cf0-4a59-aa6d-6497bf82b2cc.png)

Continuum & spectral line imaging
-----
Looks ok. There are a few issues that can be fixed later when we combine with 12m/TP.
- As previously noted, the continuum image of spw20 has a questionable feature, which may be a matter of bad line-free channel selection.
- Possible divergence in the spw24 cube:
![Brick7m_spw24](https://user-images.githubusercontent.com/3944252/175255410-37a21021-c9f0-4482-aa0c-6fc52c50319a.png)

**Comment by xinglunju** [comment]
I recleaned the SPW24 cube on a local machine with cyclefactor = 2.0. No divergence can be seen. @pyhsiehATalma please use the updated cube for image combination.

I am uploading the images to Globus at: /upload/Sgr_A_st_ao_03_7M_reclean/

I have submitted a pull request to update the json file: #357

**Comment by keflavich** [comment]
Updated files have been moved to `/orange/adamginsburg/ACES/data/2021.1.00172.L/science_goal.uid___A001_X1590_X30a8/group.uid___A001_X1590_X30a9/member.uid___A001_X15a0_X192/calibrated/working/`.  Old files are in `/orange/adamginsburg/ACES/data/2021.1.00172.L/science_goal.uid___A001_X1590_X30a8/group.uid___A001_X1590_X30a9/member.uid___A001_X15a0_X192/calibrated/working/oldspw24`

**Comment by ericliang45** [comment]
Hi there,
I was looking at the images on Globus and found multiple versions of "ALMA pipeline stage". @d-l-walker suggested I report it here so that @keflavich could remove the duplicated file as necessary. Details below:

- Navigate to folder "/rawdata/2021.1.00172.L/science_goal.uid___A001_X1590_X30a8/group.uid___A001_X1590_X30a9/member.uid___A001_X15a0_X192/calibrated/working/"
- filter by "uid___A001_*sci.spw22*cube*"
- "*s13_0*" and "*s38_0*" products co-exist, especially for cleaned cubes with "*iter1*".

ps. I didn't check how wide the scope of this multiplication is, beyond the said filtering I used.

Hope it helps.
